File: PROJEULER/p92.py

# project euler
# problem 92
# Square digit chains

mem = [0 for i in range(1000)]
sq = [i**2 for i in range(10)]
solution = [None for i in range(1000)]
lim = 1000

def sq_sum(n):
    sum = 0
    while n is not 0:
        sum += (sq[n%10])
        n//=10

    return sum

def check(n):
    global solution
    global mem
    global lim
    if n>=lim:
        n = sq_sum(n)

    if n is 89:
        return True
    elif n is 1:
        return False

    if solution[n] is None:
        solution[n] = check(n)
        return solution[n]
    else:
        return solution[n]

# def check(n):
#     global lim
#     l = [n]
#     succ = False
    
#     if n >= lim:
#         n = sq_sum(n)
#         l.append(n)

#     while n is not 89 and n is not 1:
        
#         if solution[n] is not None:
#             succ = solution[n]
#             break
#         if mem[n] is 0:
#             mem[n] = sq_sum(n)
#         n = mem[n]

#         l.append(n)
#     l.append(n)

#     if n==89:
#         succ = True

#     for i in l:
#         if i < lim:
#             solution[i] = succ

#     return succ

ans = 0
for i in range(1, 1000000):
    if check(i):
        ans += 1
    # print(i)

print(ans)

File: PROJEULER/p92.cpp
/**
* project euler
* problem 92
* Square digit chains
**/

#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <fstream>
#include <algorithm>
using namespace std;

bool solution[1000];
bool flag[1000];
int sq[10] = {0, 1, 4, 9, 16, 25, 36, 49, 64, 81};
int lim = 1000;

int sq_sum(long int n){
    int sum = 0;
    while(n!=0){
        sum += (sq[n%10]);
        n/=10;
    }

    return sum;
}

bool check(int n){
    if(n==89) return true;
    else if(n==1) return false;

    if(flag[n]){
        flag[n] = false;
        solution[n] = check(sq_sum(n));
        return solution[n];
    }else{
        return solution[n];
    }

}

bool outer_check(long int n){
    if(n>lim){
        n = sq_sum(n);
    }

    return check(n);
}

int main() {

    for(int i=0; i<lim; i++){
        flag[i] = true;
    }

    long int count=0;

    for(int i=1; i<10000000; i++){
        if(outer_check(i)) count ++;
    }

    cout<<count<<endl;

    return 0;
}
